samples/shared/SynthSound.h

```diff
@@ -65,54 +62,42 @@ class SynthSound : public IRenderableAudio {
-
-                    mPhases[j].store(mPhases[j] + mPhaseIncrements[j]);
-                    if (mPhases[j] > kTwoPi) {
-                        mPhases[j].store(mPhases[j] - kTwoPi);
-                        if (mAttackCount < kAttackPeriods)
-                        {
-                            mAttackCount++;
-                        }
-                    }
+                    audioData[i] += sinf(mPhase * (j + 1)) * mAmplitudes[j] * mMasterAmplitude;
+                }
+                mPhase.store(mPhase + mPhaseIncrement);
+                if (mPhase > kTwoPi) {
+                    mPhase.store(mPhase - kTwoPi);
                 }
             }
         } else {
             memset(audioData, 0, sizeof(float) * numFrames);
             // Remove some noise by letting the current sine wave complete.
             // Since all the frequencies are multiples of mPhases[0], if the first wave complete,
             // so should the others. Use an extra release period to reduce even more noise.
-            for (int i = 0; mPhases[0] != 0.0f
-                            && mPhases[0] < kTwoPi * (kNumExtraReleasePeriods + 1)
-                            && i < numFrames; ++i) {
+            for (int i = 0; (i < numFrames) && (mMasterAmplitude > kMasterAmplitudeCutOff); ++i) {
+                mMasterAmplitude.store(mMasterAmplitude * kReleaseMultiplier);
                 for (int j = 0; j < kNumSineWaves; ++j) {
-                    float releaseScalar = 1.0f;
-                    if (mPhases[0] > kTwoPi) {
-                        float currentPeriod = floor((mPhases[0] - kTwoPi) / kTwoPi + 1);
-                        releaseScalar -= currentPeriod / (kNumExtraReleasePeriods + 1);
-                    }
-                    audioData[i] += sinf(mPhases[j]) * mAmplitudes[j] * releaseScalar;
-                    mPhases[j].store(mPhases[j] + mPhaseIncrements[j]);
+                    audioData[i] += sinf(mPhase * (j + 1)) * mAmplitudes[j] * mMasterAmplitude;
                 }
+                mPhase.store(mPhase + mPhaseIncrement);
             }
         }
     };
 
 private:
+    std::atomic<double> mMasterAmplitude;
     std::atomic<bool> mIsWaveOn { false };
-    std::atomic<int> mAttackCount { 0 };
     std::array<std::atomic<float>, kNumSineWaves> mAmplitudes;
-    std::array<std::atomic<float>, kNumSineWaves> mPhases;
-    std::array<std::atomic<double>, kNumSineWaves> mPhaseIncrements;
-    std::array<std::atomic<double>, kNumSineWaves> mFrequencies;
+    std::atomic<float> mPhase;
+    std::atomic<double> mPhaseIncrement;
+    std::atomic<double> mFrequency { kDefaultFrequency };
     std::atomic<int32_t> mSampleRate { kDefaultSampleRate };
     void updatePhaseIncrement(){
-        for (int i = 0; i < kNumSineWaves; i++) {
-            mPhaseIncrements[i] = (kTwoPi * mFrequencies[i]) / static_cast<double>(mSampleRate);
-        }
+        mPhaseIncrement = kTwoPi * mFrequency / static_cast<double>(mSampleRate);
```

Comment:
A divide is surprisingly inefficient compared to a multiply. Cache the inverse and use that in your real-time code. This is particularly important if you a modulating the frequency.

    mSamplePeriod = 1.0 / mSampleRate;

then later

    mPhaseIncrement = ..... * mSamplePeriod;

Reply:
mPhaseIncrement is never used in any division so it's fine as is. Thanks for bringing this up though! I'll watch out for optimizations like this in the future

Reply:
> mPhaseIncrement is never used in any division so it's fine as is.

But a divide is used to calculate the mPhaseIncrement. It doesn't matter so much here because you are not varying the mPhaseIncrement. But if you add vibrato or portamento then it would be worth making this optimization.

Reply:
I added a comment for curious readers